README.md

[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![Signed
by](https://img.shields.io/badge/Keybase-Verified-brightgreen.svg)](https://keybase.io/hrbrmstr)
![Signed commit
%](https://img.shields.io/badge/Signed_Commits-100%25-lightgrey.svg)
[![Linux build
Status](https://travis-ci.org/hrbrmstr/albersusa.svg?branch=master)](https://travis-ci.org/hrbrmstr/albersusa)
[![Coverage
Status](https://codecov.io/gh/hrbrmstr/albersusa/branch/master/graph/badge.svg)](https://codecov.io/gh/hrbrmstr/albersusa)
![Minimal R
Version](https://img.shields.io/badge/R%3E%3D-3.0.0-blue.svg)
![License](https://img.shields.io/badge/License-MIT-blue.svg)

# albersusa

Tools, Shapefiles & Data to Work with an ‘AlbersUSA’ Composite
Projection

## Description

Creating a composite projection for states and counties of the United
States that includes scaled and shifted polygons for Alaska and Hawaii
is time consuming and potentially error-prone. Functions and data sets
are provided to make it easier to produce maps with a composite
projection. Furthermore, named projections for common transormations are
provided to further increase mapping productivity.

## What’s Inside The Tin

The following functions are implemented:

  - `counties_sf`: Retrieve a U.S. county composite map, optionally with
    a projection, as a simplefeature
  - `usa_sf`: Retreive a U.S. state composite map, optionally with a
    projection, as a simplefeature

New ones:

  - `points_elided`: Shift points around Alaska and Hawaii to the elided
    area (by @rdinter)

Some legacy ones:

  - `usa_composite`: Retrieve a U.S. composite map, optionally with a
    projection
  - `counties_composite`: Retrieve a U.S. county composite map,
    optionally with a projection

Pre-canned projection strings:

  - `us_aeqd_proj`: Oblique azimuthal equidistant convenience projection
  - `us_eqdc_proj`: Equidistant conic convenience projection
  - `us_laea_proj`: Albers equal-area conic convenience projection
  - `us_lcc_proj`: Lambert conformal conic convenience projection
  - `us_longlat_proj`: Generic long/lat convenience projection

The following data sets are included:

  - `system.file("extdata/composite_us_states.geojson.gz",
    package="albersusa")`
  - `system.file("extdata/composite_us_counties.geojson.gz",
    package="albersusa")`

Also, the simplefeatures columns and `@data` slot of each
`SpatialPolygonsDataFrame` has some handy data you can use (like FIPS
codes and state/county population information).

## Installation

``` r
remotes::install_git("https://git.rud.is/hrbrmstr/albersusa.git")
# or
remotes::install_git("https://git.sr.ht/~hrbrmstr/albersusa")
# or
remotes::install_gitlab("hrbrmstr/albersusa")
# or
remotes::install_bitbucket("hrbrmstr/albersusa")
# or
remotes::install_github("hrbrmstr/albersusa")
```

NOTE: To use the ‘remotes’ install options you will need to have the
[{remotes} package](https://github.com/r-lib/remotes) installed.

## Usage

``` r
library(albersusa)
library(sf)
library(sp)
library(rgeos)
library(maptools)
library(ggplot2)
library(ggalt)
library(ggthemes)
library(viridis)
library(scales)

# current version
packageVersion("albersusa")
## [1] '0.4.0'
```

### Simple features base

``` r
par(mar=c(0,0,1,0))

us_sf <- usa_sf("laea")
plot(us_sf["pop_2012"])
```

<img src="man/figures/README-unnamed-chunk-2-1.png" width="672" />

``` r

cty_sf <- counties_sf("aeqd")
plot(cty_sf["census_area"])
```

<img src="man/figures/README-unnamed-chunk-2-2.png" width="672" />

### ggplot2

``` r
ggplot() +
  geom_sf(data = us_sf, size = 0.125)
```

<img src="man/figures/README-unnamed-chunk-3-1.png" width="672" />

``` r

ggplot() +
  geom_sf(data = us_sf, size = 0.125) +
  coord_sf(crs = us_longlat_proj)
```

<img src="man/figures/README-unnamed-chunk-3-2.png" width="672" />

``` r

ggplot() +
  geom_sf(data = cty_sf, size = 0.0725)
```

<img src="man/figures/README-unnamed-chunk-3-3.png" width="672" />

``` r

ggplot() +
  geom_sf(data = cty_sf, size = 0.0725) +
  coord_sf(crs = us_longlat_proj)
```

<img src="man/figures/README-unnamed-chunk-3-4.png" width="672" />

## albersusa Metrics

| Lang | \# Files | (%) | LoC |  (%) | Blank lines |  (%) | \# Lines |  (%) |
| :--- | -------: | --: | --: | ---: | ----------: | ---: | -------: | ---: |
| R    |        9 | 0.9 | 106 | 0.77 |          66 | 0.67 |      277 | 0.86 |
| Rmd  |        1 | 0.1 |  31 | 0.23 |          32 | 0.33 |       46 | 0.14 |

## Code of Conduct

Please note that this project is released with a Contributor Code of
Conduct. By participating in this project you agree to abide by its
terms.


NEWS.md
0.4.0
* `points_elided()` works much better

0.3.0
* simplefeatures (sf) functions
* `points_elided()` function courtesy of @rdinter

0.1.0 
* Initial release


tests/test-all.R
library(testthat)
test_check("albersusa")


tests/testthat/test-albersusa.R
context("Shapefile integrity")
test_that("we can do something", {

  us <- usa_composite()
  cty <- counties_composite()

  expect_that(us, is_a("SpatialPolygonsDataFrame"))
  expect_that(cty, is_a("SpatialPolygonsDataFrame"))

  expect_that(dim(us), equals(c(51, 13)))
  expect_that(dim(cty), equals(c(3143, 8)))

  expect_that(length(us@polygons), equals(51))
  expect_that(length(cty@polygons), equals(3143))

})


R/proj.r
#' Albers equal-area conic convenience projection
#'
#' The following PROJ.4 string is used: \code{+proj=laea +lat_0=45 +lon_0=-100 +x_0=0 +y_0=0 +a=6370997 +b=6370997 +units=m +no_defs}
#'
#' In a ggplot2+ggalt context, you can just use variable name as the value for the \code{proj}
#' parameter to \code{ggalt::coord_proj()}. i.e. \code{coord_proj(us_laea_proj)}.
#'
#' @family convenience projections
#' @export
#' @examples
#' us <- usa_composite()
#' us <- sp::spTransform(us, sp::CRS(us_laea_proj))
#' us <- usa_composite(proj="laea")
#' counties <- counties_composite("laea")
us_laea_proj <- "+proj=laea +lat_0=45 +lon_0=-100 +x_0=0 +y_0=0 +a=6370997 +b=6370997 +units=m +no_defs"


#' Generic long/lat convenience projection
#'
#' The following PROJ.4 string is used: \code{+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0}
#'
#' In a ggplot2+ggalt context, you can just use variable name as the value for the \code{proj}
#' parameter to \code{ggalt::coord_proj()}. i.e. \code{coord_proj(us_longlat_proj)}.
#'
#' @export
#' @examples
#' us <- usa_composite() # it's the default
#' us <- sp::spTransform(us, sp::CRS(us_longlat_proj))
#' counties <- counties_composite()
us_longlat_proj <- "+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0"


#' Lambert conformal conic convenience projection
#'
#' The following PROJ.4 string is used: \code{+proj=lcc +lat_1=27.11637320883929 +lat_2=53.050729042644335 +lon_0=-95.44921875}
#'
#' In a ggplot2+ggalt context, you can just use variable name as the value for the \code{proj}
#' parameter to \code{ggalt::coord_proj()}. i.e. \code{coord_proj(us_lcc_proj)}.
#'
#' @family convenience projections
#' @export
#' @examples
#' us <- usa_composite()
#' us <- sp::spTransform(us, sp::CRS(us_lcc_proj))
#' us <- usa_composite(proj="lcc")
#' counties <- counties_composite("lcc")
us_lcc_proj <- "+proj=lcc +lat_1=27.11637320883929 +lat_2=53.050729042644335 +lon_0=-95.44921875"


#' Equidistant conic convenience projection
#'
#' Distance correct along meridians
#'
#' The following PROJ.4 string is used: \code{+proj=eqdc +lat_1=27.11637320883929 +lat_2=53.050729042644335 +lon_0=-95.44921875}
#'
#' In a ggplot2+ggalt context, you can just use variable name as the value for the \code{proj}
#' parameter to \code{ggalt::coord_proj()}. i.e. \code{coord_proj(us_eqdc_proj)}.
#'
#' @family convenience projections
#' @export
#' @examples
#' us <- usa_composite()
#' us <- sp::spTransform(us, sp::CRS(us_eqdc_proj))
#' us <- usa_composite(proj="eqdc")
#' counties <- counties_composite("eqdc")
#'
us_eqdc_proj <- "+proj=eqdc +lat_1=27.11637320883929 +lat_2=53.050729042644335 +lon_0=-95.44921875"


#' Oblique azimuthal equidistant convenience projection
#'
#' Distance correct along any line passing through the center of the map (i.e., great circle)
#'
#' The following PROJ.4 string is used: \code{+proj=aeqd +lat_0=40.08355112574181 +lon_0=-95.44921875}
#'
#' In a ggplot2+ggalt context, you can just use variable name as the value for the \code{proj}
#' parameter to \code{ggalt::coord_proj()}. i.e. \code{coord_proj(us_aeqd_proj)}.
#'
#' @family convenience projections
#' @export
#' @examples
#' us <- usa_composite()
#' us <- sp::spTransform(us, sp::CRS(us_aeqd_proj))
#' us <- usa_composite(proj="aeqd")
#' counties <- counties_composite("aeqd")
us_aeqd_proj <- "+proj=aeqd +lat_0=40.08355112574181 +lon_0=-95.44921875"



R/counties.r
#' Retrieve a U.S. county composite map, optionally with a projection
#'
#' The \code{proj} parameter is intende to simplify usage for those using base
#' plotting. If anything but \code{longlat} is specified the composite map
#' will be pre-projected to those
#'
#' This is what's in the SPDF \code{@data} slot:
#'
#' \preformatted{
#' Observations: 3,143
#' Variables: 9
#' $ fips        (chr) "01001", "01009", "01017", "01021", "01033", "01045", "01051",...
#' $ state_fips  (chr) "01", "01", "01", "01", "01", "01", "01", "01", "01", "01", "0...
#' $ county_fips (chr) "001", "009", "017", "021", "033", "045", "051", "065", "079",...
#' $ name        (chr) "Autauga", "Blount", "Chambers", "Chilton", "Colbert", "Dale",...
#' $ lsad        (chr) "County", "County", "County", "County", "County", "County", "C...
#' $ census_area (dbl) 594.436, 644.776, 596.531, 692.854, 592.619, 561.150, 618.485,...
#' $ state       (chr) "Alabama", "Alabama", "Alabama", "Alabama", "Alabama", "Alabam...
#' $ iso_3166_2  (chr) "AL", "AL", "AL", "AL", "AL", "AL", "AL", "AL", "AL", "AL", "A...
#' $ population  (dbl) 54590, 57466, 34228, 43611, 54451, 50129, 79330, 15750, 34148,...
#' }
#'
#' A reference GeoJSON file isprovided in the following location (i.e. you can use
#' it in any mapping program):
#'
#' \preformatted{
#' system.file("extdata/composite_us_counties.geojson.gz", package="albersusa")
#' }
#'
#' @param proj if anything but \code{longlat} is specified, the shapefile will
#'        be pre-projected before being returned
#' @export
#' @examples
#' library(sp)
#'
#' plot(counties_composite())
#' plot(counties_composite("laea"))
counties_composite <- function(proj=c("longlat", "laea", "lcc", "eqdc", "aeqd")) {

  # us <- readOGR(system.file("extdata/composite_us_counties.geojson", package="albersusa"),
  #               "OGRGeoJSON",
  #               stringsAsFactors=FALSE,
  #               verbose=FALSE)

  cty <- readRDS(system.file("extdata/counties.rda", package="albersusa"))

  proj <- match.arg(proj, c("longlat", "laea", "lcc", "eqdc", "aeqd"))

  if (proj != "longlat") {

    proj <- switch(proj,
                   laea=us_laea_proj,
                   lcc=us_lcc_proj,
                   eqdc=us_eqdc_proj,
                   aeqd=us_aeqd_proj)

    cty <- sp::spTransform(cty, sp::CRS(proj))

  }

  cty

}


R/albersusa.R
#' Retrieve a U.S. state composite map, optionally with a projection
#'
#' The \code{proj} parameter is intende to simplify usage for those using base
#' plotting. If anything but \code{longlat} is specified the composite map
#' will be pre-projected to those
#'
#' This is what's in the SPDF \code{@data} slot:
#'
#' \preformatted{
#' Observations: 51
#' Variables: 13
#' $ geo_id              (chr) "0400000US04", "0400000US05", "0400000US06", "0400...
#' $ fips_state          (chr) "04", "05", "06", "08", "09", "11", "13", "17", "1...
#' $ name                (chr) "Arizona", "Arkansas", "California", "Colorado", "...
#' $ iso_3166_2          (chr) "AZ", "AR", "CA", "CO", "CT", "DC", "GA", "IL", "I...
#' $ census              (int) 6392017, 2915918, 37253956, 5029196, 3574097, 6017...
#' $ pop_estimataes_base (int) 6392310, 2915958, 37254503, 5029324, 3574096, 6017...
#' $ pop_2010            (int) 6411999, 2922297, 37336011, 5048575, 3579345, 6052...
#' $ pop_2011            (int) 6472867, 2938430, 37701901, 5119661, 3590537, 6204...
#' $ pop_2012            (int) 6556236, 2949300, 38062780, 5191709, 3594362, 6350...
#' $ pop_2013            (int) 6634997, 2958765, 38431393, 5272086, 3599341, 6491...
#' $ pop_2014            (int) 6731484, 2966369, 38802500, 5355866, 3596677, 6588...
#' }
#'
#' A reference GeoJSON file is provided in the following location (i.e. you can use
#' it in any mapping program):
#'
#' \preformatted{
#' system.file("extdata/composite_us_states.geojson.gz", package="albersusa")
#' }
#'
#' @param proj if anything but \code{longlat} is specified, the shapefile will
#'        be pre-projected before being returned
#' @export
#' @examples
#' library(sp)
#'
#' plot(usa_composite())
#' plot(usa_composite("laea"))
usa_composite <- function(proj=c("longlat", "laea", "lcc", "eqdc", "aeqd")) {

  # us <- readOGR(system.file("extdata/composite_us_states.geojson", package="albersusa"),
  #               "OGRGeoJSON",
  #               stringsAsFactors=FALSE,
  #               verbose=FALSE)

  us <- readRDS(system.file("extdata/states.rda", package="albersusa"))

  proj <- match.arg(proj, c("longlat", "laea", "lcc", "eqdc", "aeqd"))

  if (proj != "longlat") {

    proj <- switch(proj,
                   laea=us_laea_proj,
                   lcc=us_lcc_proj,
                   eqdc=us_eqdc_proj,
                   aeqd=us_aeqd_proj)

    us <- sp::spTransform(us, sp::CRS(proj))

  }

  us

}

R/counties_sf.r
#' Retrieve a U.S. county composite map, optionally with a projection, as a simplefeature
#'
#' The \code{proj} parameter is intende to simplify usage for those using base
#' plotting. If anything but \code{longlat} is specified the composite map
#' will be pre-projected to those
#'
#' This is what's in the object looks like:
#'
#' \preformatted{
#' Observations: 3,143
#' Variables: 9
#' $ fips        <fctr> 01001, 01009, 01017, 01021, 01033, 01045, 01051, 01065, 01079, 0...
#' $ state_fips  <fctr> 01, 01, 01, 01, 01, 01, 01, 01, 01, 01, 01, 01, 01, 05, 05, 06, ...
#' $ county_fips <fctr> 001, 009, 017, 021, 033, 045, 051, 065, 079, 083, 099, 107, 121,...
#' $ name        <fctr> Autauga, Blount, Chambers, Chilton, Colbert, Dale, Elmore, Hale,...
#' $ lsad        <fctr> County, County, County, County, County, County, County, County, ...
#' $ census_area <dbl> 594.436, 644.776, 596.531, 692.854, 592.619, 561.150, 618.485, 64...
#' $ state       <fctr> Alabama, Alabama, Alabama, Alabama, Alabama, Alabama, Alabama, A...
#' $ iso_3166_2  <fctr> AL, AL, AL, AL, AL, AL, AL, AL, AL, AL, AL, AL, AL, AR, AR, CA, ...
#' $ geometry    <simple_feature> MULTIPOLYGON(((-86.496774 3..., MULTIPOLYGON(((-86.577...#' }
#'
#' A reference GeoJSON file isprovided in the following location (i.e. you can use
#' it in any mapping program):
#'
#' \preformatted{
#' system.file("extdata/composite_us_counties.geojson.gz", package="albersusa")
#' }
#'
#' @param proj if anything but \code{longlat} is specified, the shapefile will
#'        be pre-projected before being returned
#' @export
#' @examples
#' library(sf)
#'
#' plot(counties_sf()) # these take a while to render
#' plot(counties_sf("laea"))
counties_sf <- function(proj=c("longlat", "laea", "lcc", "eqdc", "aeqd")) {

  # us <- readOGR(system.file("extdata/composite_us_counties.geojson", package="albersusa"),
  #               "OGRGeoJSON",
  #               stringsAsFactors=FALSE,
  #               verbose=FALSE)

  cty <- readRDS(system.file("extdata/counties_sf.rda", package="albersusa"))

  proj <- match.arg(proj, c("longlat", "laea", "lcc", "eqdc", "aeqd"))

  if (proj != "longlat") {

    proj <- switch(proj,
                   laea=us_laea_proj,
                   lcc=us_lcc_proj,
                   eqdc=us_eqdc_proj,
                   aeqd=us_aeqd_proj)

    cty <- sf::st_transform(cty, proj)

  }

  cty

}


R/points_elided.R
#' Shift points around Alaska and Hawaii to the elided area
#'
#' This function will take a SpatialPoints object or a data frame of coordinates
#' and shift the points around Alaska and Hawaii to the elided area from this package,
#' leaving the other points intact.
#'
#' @param sp An object of SpatialPoints class or a data frame with x (`lon`) and y (`lat`)
#' @return An elided version of the original SpatialPoints class or a data frame
#'         depending on what was passed in.
#' @export
points_elided <- function(sp) {

  ret <- "sp"

  if (inherits(sp, "data.frame")) {
    class(sp) <- "data.frame"
    sp <- setNames(sp, c("lon", "lat"))
    sp::coordinates(sp) <- ~lon+lat
    sp::proj4string(sp) <- us_longlat_proj
    ret <- "df"
  }

  orig_proj <- sp::proj4string(sp)

  # convert it to Albers equal area
  sp <- sp::spTransform(sp, sp::CRS("+proj=laea +lat_0=45 +lon_0=-100 +x_0=0 +y_0=0 +a=6370997 +b=6370997 +units=m +no_defs"))

  ak_bb <- readRDS(system.file("extdata/alaska_bb.rda", package="albersusa"))
  ak_poly <- as(raster::extent(as.vector(t(ak_bb))), "SpatialPolygons")
  sp::proj4string(ak_poly) <- "+proj=laea +lat_0=45 +lon_0=-100 +x_0=0 +y_0=0 +a=6370997 +b=6370997 +units=m +no_defs"

  hi_bb <- readRDS(system.file("extdata/hawaii_bb.rda", package="albersusa"))
  hi_poly <- as(raster::extent(as.vector(t(hi_bb))), "SpatialPolygons")
  sp::proj4string(hi_poly) <- "+proj=laea +lat_0=45 +lon_0=-100 +x_0=0 +y_0=0 +a=6370997 +b=6370997 +units=m +no_defs"

  spl <- vector("list", length(sp))

  for (idx in seq_along(sp)) {

    tmp <- sp[idx]

    if (!is.na(sp::over(tmp, ak_poly))) {
      tmp <- maptools::elide(tmp, scale = max(apply(ak_bb, 1, diff)) / 2.3, rotate = -50, bb = ak_bb)
      tmp <- maptools::elide(tmp, shift = c(-1298669, -3018809))
    } else if (!is.na(sp::over(tmp, hi_poly))) {
      tmp <- maptools::elide(tmp, rotate = -35, bb = hi_bb)
      tmp <- maptools::elide(tmp, shift = c(5400000, -1400000))
    }

    spl[[idx]] <- sp::coordinates(tmp)

  }

  sp <- do.call(rbind, spl)
  rownames(sp) <- 1:nrow(sp)
  sp <- sp::SpatialPoints(sp, proj4string = sp::CRS(orig_proj))

  return(if (ret == "sp") sp else as.data.frame(sp))

}


R/albersusa-package.R
#' Tools, Shapefiles & Data to Work with an 'AlbersUSA' Composite Projection
#'
#' Creating a composite projection for states and counties of the United States
#' that includes scaled and shifted polygons for Alaska and Hawaii is time consuming
#' and potentially error-prone. Functions and data sets are provided to make it easier to
#' produce maps with a composite projection. Furthermore, named projections for common
#' transormations are provided to further increase mapping productivity.
#'
#' Newer \code{sf} (simple features) objects can be returned or legacy R
#' spatial \code{sp} objects can be returned.
#'
#' Reference GeoJSON files are provided in the following locations (i.e. you can use
#' them in any mapping program):
#'
#' \preformatted{
#' system.file("extdata/composite_us_states.geojson.gz", package="albersusa")
#' system.file("extdata/composite_us_counties.geojson.gz", package="albersusa")
#' }
#'
#' @name albersusa
#' @docType package
#' @author Bob Rudis (bob@@rud.is)
#' @import sp rgeos rgdal maptools
#' @import sf
NULL


R/states_sf.r
#' Retreive a U.S. state composite map, optionally with a projection, as a simplefeature
#'
#' The \code{proj} parameter is intende to simplify usage for those using base
#' plotting. If anything but \code{longlat} is specified the composite map
#' will be pre-projected to those
#'
#' This is what's the object looks like:
#'
#' \preformatted{
#' Observations: 51
#' Variables: 14
#' $ geo_id              <fctr> 0400000US04, 0400000US05, 0400000US06, 0400000US08, 0400...
#' $ fips_state          <fctr> 04, 05, 06, 08, 09, 11, 13, 17, 18, 22, 27, 28, 30, 35, ...
#' $ name                <fctr> Arizona, Arkansas, California, Colorado, Connecticut, Di...
#' $ lsad                <fctr> , , , , , , , , , , , , , , , , , , , , , , , , , , , , ,
#' $ census_area         <dbl> 113594.084, 52035.477, 155779.220, 103641.888, 4842.355, ...
#' $ iso_3166_2          <fctr> AZ, AR, CA, CO, CT, DC, GA, IL, IN, LA, MN, MS, MT, NM, ...
#' $ census              <int> 6392017, 2915918, 37253956, 5029196, 3574097, 601723, 968...
#' $ pop_estimataes_base <int> 6392310, 2915958, 37254503, 5029324, 3574096, 601767, 968...
#' $ pop_2010            <int> 6411999, 2922297, 37336011, 5048575, 3579345, 605210, 971...
#' $ pop_2011            <int> 6472867, 2938430, 37701901, 5119661, 3590537, 620427, 981...
#' $ pop_2012            <int> 6556236, 2949300, 38062780, 5191709, 3594362, 635040, 991...
#' $ pop_2013            <int> 6634997, 2958765, 38431393, 5272086, 3599341, 649111, 999...
#' $ pop_2014            <int> 6731484, 2966369, 38802500, 5355866, 3596677, 658893, 100...
#' $ geometry            <list> [-112.53859, -112.53454, -112.36895, -112.35769, -111.41...#' }
#'
#' A reference GeoJSON file is provided in the following location (i.e. you can use
#' it in any mapping program):
#'
#' \preformatted{
#' system.file("extdata/composite_us_states.geojson.gz", package="albersusa")
#' }
#'
#' @param proj if anything but \code{longlat} is specified, the shapefile will
#'        be pre-projected before being returned
#' @export
#' @examples
#' library(sf)
#'
#' plot(usa_sf()) # these take a while to render
#' plot(usa_sf("laea"))
usa_sf <- function(proj=c("longlat", "laea", "lcc", "eqdc", "aeqd")) {

  # us <- readOGR(system.file("extdata/composite_us_states.geojson", package="albersusa"),
  #               "OGRGeoJSON",
  #               stringsAsFactors=FALSE,
  #               verbose=FALSE)

  us <- readRDS(system.file("extdata/states_sf.rda", package="albersusa"))

  proj <- match.arg(proj, c("longlat", "laea", "lcc", "eqdc", "aeqd"))

  if (proj != "longlat") {

    proj <- switch(proj,
                   laea=us_laea_proj,
                   lcc=us_lcc_proj,
                   eqdc=us_eqdc_proj,
                   aeqd=us_aeqd_proj)

    us <- sf::st_transform(us, proj)

  }

  us

}